### PR TITLE
Add gpt model, and fix gpt dataset bug

### PR DESCRIPTION
**Pull request by marigoold**
此pr完成了：
- 增加gpt模型
- 修复了gpt dataset调用helpers.so会越界报错的bug

### CODE REVIEW

File: configs/gpt_large_pretrain.py [hunk 1]
```diff
@@ -7,10 +7,19 @@
 from .common.models.graph import graph
 
 # Bert-large model config
```
Comment (L1aoXingyu) [inline configs/gpt_large_pretrain.py:9]: 这里注释改一下，然后文件名也改一下，gpt 的命名应该没有 gpt_large 这种说法，应该是 GPT-2 之类的

File: configs/gpt_large_pretrain.py [hunk 2]
```diff
@@ -7,10 +7,19 @@
...
 
-train.train_micro_batch_size = 16
+for ds in dataloader.train.dataset:
+    ds.max_seq_length = model.cfg.max_seq_length
+
+optim.lr = 1.5e-4
+
+train.train_micro_batch_size = 4
 train.recompute_grad.enabled = True
 train.output_dir = "./demo_output"
```
Comment (L1aoXingyu) [inline configs/gpt_large_pretrain.py:25]: 这里改一下，默认放在 output 下面，可以是 `output/gpt_2` 之类的

File: libai/models/gpt_model.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+
+    def __init__(self, max_positions=1024, *, layer_idx=0):
+        super().__init__()
+        self.mask = flow.tril(
+            flow.ones(
+                (max_positions, max_positions),
+                dtype=flow.int8,
+                placement=dist.get_layer_placement(layer_idx),
+                sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
+                # sbp=[flow.sbp.broadcast, flow.sbp.broadcast]
```
Comment (L1aoXingyu) [inline libai/models/gpt_model.py:49]: 注释删掉

File: libai/models/gpt_model.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        layernorm_epsilon=1e-5,
+        initializer_range=0.02,
+        use_scaled_init_for_output_weights=True,
+        bias_gelu_fusion=False,
+        bias_dropout_fusion=False,
+        scale_mask_softmax_fusion=False,
+        apply_query_key_layer_scaling=False,
+    ):
+        super().__init__()
+        # init_method = init_method_normal(std=initializer_range)
```
Comment (L1aoXingyu) [inline libai/models/gpt_model.py:113]: 注释删掉

File: libai/models/gpt_model.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        vocab_size,
+        hidden_size,
+        max_seq_length,
+        init_method=init.xavier_normal_,
+        embedding_dropout_prob=0.0,
+    ):
+        super().__init__()
+        self.token_embeddings = VocabEmbedding(vocab_size, hidden_size, init_method=init_method)
+        self.position_embeddings = Embedding(max_seq_length, hidden_size, init_method=init_method)
+        self.dropout = flow.nn.Dropout(embedding_dropout_prob)
```
Comment (L1aoXingyu) [inline libai/models/gpt_model.py:193]:
```suggestion
        self.dropout = nn.Dropout(embedding_dropout_prob)
```

File: libai/models/gpt_model.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        position_ids = position_ids.expand_as(input_ids).to_global(sbp=input_ids.sbp)
+
+        token_embeds = self.token_embeddings(input_ids)
+        position_embeds = self.position_embeddings(position_ids)
+        input_embeds = token_embeds + position_embeds
+        input_embeds = self.dropout(input_embeds)
+        return input_embeds
+
+
+class Transformer(nn.Module):
```
Comment (L1aoXingyu) [inline libai/models/gpt_model.py:215]: 感觉可以把这个 module 直接写到 GPTModel 里面

**Comment by L1aoXingyu** [comment]
增加一个 gpt 模型的 unittest

File: libai/data/datasets/gpt_dataset.py [hunk 7]
```diff
@@ -35,10 +35,10 @@ def __len__(self):
     def __getitem__(self, idx):
         text = np.array(self.dataset[idx], dtype=np.long)
         input_ids = flow.tensor(text[:-1], dtype=flow.long)
-        labels = flow.tensor(text[1:], dtype=flow.long)
+        lm_labels = flow.tensor(text[1:], dtype=flow.long)
         sample = Instance(
             input_ids=DistTensorData(input_ids),
-            labels=DistTensorData(labels, placement_idx=-1),
+            lm_labels=DistTensorData(lm_labels, placement_idx=-1),
```
Comment (CPFLAME) [inline libai/data/datasets/gpt_dataset.py:41]: 这里可能得注意一下,  evaluator里面默认读取的字段是`labels`
但如果gpt没有用到evaluator的话, 可以不管

File: libai/models/gpt_model.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        self.loss_func = GPTLoss()
+
+    def forward(
+        self,
+        input_ids,
+        lm_labels=None,
+    ):
+        logits = self.GPT_model(input_ids)
+        lm_loss = self.loss_func(logits, lm_labels)
+        return lm_loss
```
Comment (CPFLAME) [inline libai/models/gpt_model.py:291]: 这里可以和T5一样,  加个判断
https://github.com/Oneflow-Inc/libai/blob/5a3c353d093fa25570e7a8a74303e70beb4c0de7/libai/models/t5_model.py#L276-L282

File: configs/gpt_large_pretrain.py (hunk 2 above)
Comment (marigoold) [reply configs/gpt_large_pretrain.py]: 已改

File: configs/gpt_large_pretrain.py (hunk 1 above)
Comment (marigoold) [reply configs/gpt_large_pretrain.py]: 已改

File: libai/models/gpt_model.py (hunk 3 above)
Comment (marigoold) [reply libai/models/gpt_model.py]: 已改

File: libai/models/gpt_model.py (hunk 4 above)
Comment (marigoold) [reply libai/models/gpt_model.py]: 已改

File: libai/models/gpt_model.py (hunk 5 above)
Comment (marigoold) [reply libai/models/gpt_model.py]: 已改

**Comment by marigoold** [comment]
> 增加一个 gpt 模型的 unittest

已增加

File: libai/models/gpt_model.py (hunk 8 above)
Comment (marigoold) [reply libai/models/gpt_model.py]: 已改

File: libai/data/datasets/gpt_dataset.py (hunk 7 above)
Comment (marigoold) [reply libai/data/datasets/gpt_dataset.py]: 已改，模型里面也改了

**Comment by marigoold** [comment]
helpers根据dangkai的建议改了，在GPT上面没问题，但是还没验证在t5上能不能跑通

**Comment by marigoold** [comment]
> > helpers的c++扩展根据dangkai的建议改了，在GPT上面没问题，但是还没验证在t5上能不能跑通
> 
> 可以验证一下

验证过了，T5模型没用到c++扩展改过的函数，所以没有受到影响

**Review by CPFLAME — APPROVED** [review]

**Review by L1aoXingyu — APPROVED** [review]
